### ISSUE TITLE
Button label or slot

### ISSUE THREAD

**Issue by timothyjoh** [material-components/material-components-web-components#256]
Now we have 2 ways of applying text into the button, one via the label, and the other the slot.

The leading and trailing icons work fine around the label property, but the slot (if you put anything in there) breaks the styling.

Seems to me we should remove either the label prop, or the slot, both with consequences:
1. Using the slot (but rendering it inside the mdc-button__label span) would keep the API more similar to the HTML button tag, and be more familiar
2. Using the label only (eliminating the slot) would force the incoming text to be text only, and no nesting of other elements.

Any thoughts from the others?

**Comment by Westbrook** [comment]
This may be more tricky that it needs to be, but you can apply via either approach if you update:
```
        <span class="mdc-button__label">${this.label}</span>
        ${this.icon && this.trailingIcon ? mdcButtonIcon : ''}
        <slot></slot>
```
To be:
```
        <span class="mdc-button__label"><slot>${this.label}</slot></span>
        ${this.icon && this.trailingIcon ? mdcButtonIcon : ''}
```
Which would prefer the slotted content for the button but fallback to the `label` value when that isn't present allowing both APIs to have the same qualitative effects on the element.

The only negative side effect here is the `<mwc-button label="Text"> </mwc-button>` would be rendered with no text due to the `[space]` character in the slotted text node. Something that could be documented and avoided, but can be pretty tricky when you first encounter it in usage.

**Comment by jhnsntmthy** [comment]
That is why I feel that `slot` should be the one used here, and not the `label` prop

**Comment by elyobo** [comment]
:+1: for `slot` and dropping `label` entirely - `slot` is more flexible than `label`, it's consistent with `<button>`, it avoids weird spacing gotchas as pointed out by @Westbrook, and it means there's only one way to do something (more simple API) than trying to support both.

PR welcome?

**Comment by azakus** [comment]
It looks like the `label` property both sets the label text, _and_ sets `aria-label`. We probably need to verify that a technique using `slot` by default will work correctly.

**Comment by asyncLiz** [comment]
Per spec, labels can only contain text and not arbitrary HTML content. Therefore we want to avoid styling problems by using a slot.

The recommendation is to use the `label` property. We'll be researching extensibility in #602, which would allow you to extend the button more easily and change the content of the label to include things other than text.
